crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -1356,24 +1363,24 @@ impl CheckpointSignatureAggregator {
         let their_digest = *data.summary.digest();
         let (_, signature) = data.summary.into_data_and_sig();
         let author = signature.authority;
-        // It is not guaranteed that signature.authority == narwhal_cert.author, but we do verify
-        // the signature so we know that the author signed the message at some point.
-        if their_digest != self.digest {
-            self.metrics.remote_checkpoint_forks.inc();
-            warn!(
-                checkpoint_seq = self.summary.sequence_number,
-                "Validator {:?} has mismatching checkpoint digest {}, we have digest {}",
-                author.concise(),
-                their_digest,
-                self.digest
-            );
-            return Err(());
-        }
-
         let envelope =
             SignedCheckpointSummary::new_from_data_and_sig(self.summary.clone(), signature);
-        match self.signatures.insert(envelope) {
-            InsertResult::Failed { error } => {
+        match (
+            self.signatures.insert(envelope),
```

Comment:
This now means we are unconditionally insert this even if their digest could be different from ours. This is a bit confusing and I have a feeling maybe you could simply these two fields by replacing them with only one, the voting_history.

Reply:
> This now means we are unconditionally insert this even if their digest could be different from ours.

Well we can't return early, as if we return early once we see we disagree, then the condition will go unnoticed. But I see now that we may never detect that we disagree unless it's a split brain with this. I think I can just use voting_history to detect this only before certifying instead.

> This is a bit confusing and I have a feeling maybe you could simply these two fields by replacing them with only one, the voting_history

I don't think we can. `voting_history` intentionally has a weak quorum requirement, because we're trying to tally up to f+1 for two different sets to determine split brain.

Reply:
I can create a single data structure that encapsulates the complexity.. I'm assuming that is what you meant

Reply:
Ah I see the problem. We may never notice the digest mismatch